agora/Agora/Utils.hs

```diff
@@ -326,37 +338,126 @@ ptokenSpent =
 {- | True if both maps have exactly the same keys.
      Using @'#=='@ is not sufficient, because keys returned are not ordered.
 -}
-pkeysEqual :: forall (s :: S) k a b. Term s (PMap k a :--> PMap k b :--> PBool)
+pkeysEqual :: (POrd k, PIsData k) => forall (s :: S) a b. Term s (PMap k a :--> PMap k b :--> PBool)
 pkeysEqual = phoistAcyclic $
   plam $ \p q -> P.do
     pks <- plet $ pkeys # p
     qks <- plet $ pkeys # q
-    pall # plam (\pk -> pelem # pk # qks) # pks
-      #&& pall # plam (\qk -> pelem # qk # pks) # qks
 
--- | / O(n^2) /. Clear out duplicates in a list. The order is not preserved.
-pnub :: forall list a (s :: S). (PEq a, PIsListLike list a) => Term s (list a :--> list a)
-pnub =
-  phoistAcyclic $
-    precList
-      ( \self x xs ->
+    pif
+      (plength # pks #== plength # qks)
+      ( P.do
+          let comp = phoistAcyclic $ plam $ \(pfromData -> x) (pfromData -> y) -> x #< y
+              spks = pmsort # comp # pks
+              sqks = pmsort # comp # qks
+
+          plistEquals # spks # sqks
+      )
+      (pcon PFalse)
+
+-- | / O(nlogn) /. Clear out duplicates in a list. The order is not preserved.
```

Comment:
Did you write these util functions yourself?

Reply:
Yes.